internal/feature.go

```diff
@@ -47,54 +57,74 @@ type featureTest struct {
 //	err != nil: the test couldn't be executed
 type FeatureTestFn func() error
 
-// FeatureTest wraps a function so that it is run at most once.
-//
-// name should identify the tested feature, while version must be in the
-// form Major.Minor[.Patch].
+// NewFeatureTest is a convenient way to create a single [FeatureTest].
+func NewFeatureTest(name, version string, fn FeatureTestFn) func() error {
+	return (&FeatureTest{
+		Name:    name,
+		Version: version,
+		Fn:      fn,
+	}).Result
+}
+
+// Result returns the outcome of a feature test, executing it if necessary.
 //
-// Returns an error wrapping ErrNotSupported if the feature is not supported.
-func FeatureTest(name, version string, fn FeatureTestFn) func() error {
-	ft := new(featureTest)
-	return func() error {
-		ft.RLock()
-		if ft.successful {
-			defer ft.RUnlock()
-			return ft.result
-		}
-		ft.RUnlock()
-		ft.Lock()
-		defer ft.Unlock()
-		// check one more time on the off
-		// chance that two go routines
-		// were able to call into the write
-		// lock
-		if ft.successful {
-			return ft.result
+// See [FeatureTestFn] for the meaning of the returned error.
+func (ft *FeatureTest) Result() error {
+	ft.mu.RLock()
+	if ft.successful {
+		defer ft.mu.RUnlock()
+		return ft.result
+	}
+	ft.mu.RUnlock()
+	ft.mu.Lock()
+	defer ft.mu.Unlock()
+	// check one more time on the off
+	// chance that two go routines
+	// were able to call into the write
+	// lock
+	if ft.successful {
+		return ft.result
+	}
+	err := ft.Fn()
+	switch {
```

Comment:
This is still too contrived (I know it's old code). A more common pattern:

```go
if err := ft.Fn(); !errors.Is(err, ErrNotSupported) {
  return fmt.Errorf()
}

ft.done = true
ft.result = &UFE{}

return ft.result
```

Reply:
It's  not so straight forward because you also have to deal with the `err == nil` case which should create UFE. I changed it into an if, else, not sure it's much more readable tbh.

Reply:
```go
	err := ft.Fn()
	if err == nil {
		ft.done = true
		return nil
	}
	if errors.Is(err, ErrNotSupported) {
		ft.done = true
		ft.result = &UnsupportedFeatureError{
			// MinimumVersion: v,
			Name: ft.Name,
		}

		return ft.result
	}

	// We couldn't execute the feature test to a point
	// where it could make a determination.
	// Don't cache the result, just return it.
	return fmt.Errorf("detect support for %s: %w", ft.Name, err)
}
```